### PR TITLE
Add support for locations starting with `https://github.com/`

### PR DESCRIPTION
**Pull request by mikelik**
Additionally get rid of `gh` error if not installed.
Fix an update of the existing dependency: previously update was ignored (because executing `emplace` on `unordered_map` is ignored if `key` already exists).
When it comes to supporting `https://github.com`: now dependencies are still stored as shorthands (i.e. `org/project`), so I'm normalizing the location before validating / adding location to the dependency.

### CODE REVIEW

File: include/antler/project/object.hpp [hunk 1]
```diff
@@ -95,7 +95,11 @@ namespace antler::project {
          const auto& itr = m_dependencies.find(dep.name());
          bool has_value = itr == m_dependencies.end();
 
-         m_dependencies.emplace(dep.name(), std::move(dep));
+         if(!has_value)
+            itr->second = dep;
+         else
+            m_dependencies.emplace(dep.name(), std::move(dep));
```
Comment (mikelik) [inline include/antler/project/object.hpp:101]: Is this undefined behavior because we don't know what will be the order of evaluating arguments?
So first we might move `dep` and later read `dep.name()` which will be undefined?

File: src/location.cpp [hunk 2]
```diff
@@ -32,14 +42,21 @@ bool is_github_shorthand(std::string_view s) { return github::is_shorthand(s); }
 
 bool is_github_repo(std::string_view s) { return is_github(s) && !is_archive(s); }
 
+std::string_view strip_github_url_to_shorthand(std::string_view github_url) {
+   if(github_url.size() <= github_com.size())
+      throw std::out_of_range("URL passed: "+  std::string(github_url) + " should have minimum size of " + std::to_string(github_com.size() + 1));
```
Comment (ScottBailey) [inline src/location.cpp:47]: More specific sanity checking? 
```suggestion
   if(github_url.find(github_com) != 0)
        throw std::out_of_range("URL passed: "+  std::string(github_url) + " should begin with: " + github_com);
```

File: include/antler/project/object.hpp (hunk 1 above)
Comment (ScottBailey) [reply include/antler/project/object.hpp:101]: If this is undefined behavior, I think it's more like forwarding happens on name and move before construction of the key, val pair. So when the pair is created (the internal storage for unordered_map) name is no longer valid.

Did you try this with [UBsan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#id3)?

**Review by ScottBailey — CHANGES_REQUESTED** [review]
Looks pretty good to me. I added a couple things to look at.

File: include/antler/project/object.hpp (hunk 1 above)
Comment (dimas1185) [reply include/antler/project/object.hpp:101]: to my best knowledge this is UB as order of arguments evaluation of function is unspecified. So even if such thing work I wouldn't rely on it as on some compiler (or specific version of it) it could break.
If you want to deeper understand that, google for 'order of evaluation' and 'sequence points'. I know there are some changes in c++17 comparing to older standards.

File: include/antler/project/object.hpp [hunk 3]
```diff
@@ -95,7 +95,11 @@ namespace antler::project {
          const auto& itr = m_dependencies.find(dep.name());
          bool has_value = itr == m_dependencies.end();
```
Comment (dimas1185) [inline include/antler/project/object.hpp:96]: sounds more logical and if you do this please update accordingly logic below.
```suggestion
         bool has_value = itr != m_dependencies.end();
```

File: include/antler/system/utils.hpp [hunk 4]
```diff
@@ -86,6 +86,7 @@ namespace antler::system {
       for (const auto& arg : args) {
          cmd += " " + arg;
       }
+      cmd += " 2>&1";
```
Comment (dimas1185) [inline include/antler/system/utils.hpp:89]: what does it affect? if there will be error you return `nullopt` and if not you'll unlikely to have error output at all. And even if you do, parsing of output will likely break.

File: src/location.cpp [hunk 5]
```diff
@@ -22,7 +22,17 @@ bool is_archive(std::string_view s) {
          ends_with(s, ".tar.zst");
 }
 
-static inline bool is_github(std::string_view s) { return starts_with(s, "https://github.com"); }
+constexpr std::string_view github_com = "https://github.com/";
+
+std::string_view normalize(std::string_view loc) {
+   if (is_github_repo(loc)) {
+      return strip_github_url_to_shorthand(loc);
```
Comment (dimas1185) [inline src/location.cpp:29]: I suggest to combine it with `is_github_repo` or refactor in other way. Currently this function doesn't check for `github.com` to be at the beginning. That means that its call will break a string if called without `is_github_repo`.
That is error prone. I would either call `is_github_repo` inside `strip_github_url_to_shorthand` or make a generic `strip_host_from_url` function that will be called like that: `strip_host_from_url(github_com, loc)`

File: include/antler/project/location.hpp [hunk 6]
```diff
@@ -9,13 +9,21 @@
 
 namespace antler::project::location {
 
+/// If location is github URL then strips out `https://github.com/` from URL
+/// @param loc Github location - either full URL or shorthand
+/// @return string_view shorthand version of full github URL
+[[nodiscard]] std::string_view normalize(std::string_view loc);
```
Comment (dimas1185) [inline include/antler/project/location.hpp:15]: I find name `normalize` too ambiguous. I know this name is something used for unicode conversion or other representation format like adding trailing zeros to a number, etc. I would name it `relative_path` or something.

File: tests/dependency_tests.cpp [hunk 7]
```diff
@@ -18,13 +18,6 @@ TEST_CASE("Testing dependency validating locations") {
    CHECK(d.location() == "larryk85/does-not-exist");
 
    CHECK(!d.is_valid_location());
-
-   // TODO: reinstate these tests when support for general git repos and archives is added
-   //d.location("https://github.com/larryk85/cturtle");
```
Comment (dimas1185) [inline tests/dependency_tests.cpp:23]: why did you remove it? It sounds like the time to uncomment it.

**Review by dimas1185 — CHANGES_REQUESTED** [review]
see my comments

File: include/antler/project/object.hpp (hunk 1 above)
Comment (mikelik) [reply include/antler/project/object.hpp:101]: > Did you try [UBsan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#id3)?
> 
> I think the process of emplace will do a forward of both args into the next function at essentially the same time. Not really an order of evaluation.

I couldn't compile whole code with clang++ (due to external projects), so I have extracted following code:
https://godbolt.org/z/4953rEfh5
and UBsan did not show anything.

Later I found this: https://stackoverflow.com/questions/27266755/is-it-safe-to-pass-as-argument-a-member-of-an-object-which-is-moving

and in short: in this case it is defined.
Anyway I will change it so no one else needs to think of it. New proposed version:

```suggestion
            auto name = dep.name();
            m_dependencies.emplace(std::move(name), std::move(dep));
```

File: include/antler/system/utils.hpp (hunk 4 above)
Comment (mikelik) [reply include/antler/system/utils.hpp:89]: It resolves issue: `sh: 1: gh: not found` in the original issue.

I believe Bucky's intention was to run command line `gh --version` quietly (the function name is `execute_quiet`).
While stdout was silenced, the stderr was still printing error (which was ignored anyway).

File: tests/dependency_tests.cpp (hunk 7 above)
Comment (mikelik) [reply tests/dependency_tests.cpp:23]: Original idea lost its sense, because in location we are storing short / relative version i.e. `org/project`.
Only tests that would be valid in this file are already done above and they are testing short version.

The tests that are testing full github URL are added to location_tests.cpp and net_utils_tests.cpp

File: include/antler/project/location.hpp (hunk 6 above)
Comment (mikelik) [reply include/antler/project/location.hpp]: I have renamed it to `strip_github_com`

File: src/location.cpp (hunk 5 above)
Comment (mikelik) [reply src/location.cpp]: Combined `normalize` and `strip_github_url_to_shorthand` and this function is called now `strip_github_com`

File: include/antler/project/object.hpp (hunk 3 above)
Comment (mikelik) [reply include/antler/project/object.hpp]: done

[the remark this answers is not in the document]
Comment (mikelik) [reply include/antler/project/object.hpp]: I guess that Bucky meant `has_value_inserted`?
But anyway I have changed it

File: src/location.cpp (hunk 2 above)
Comment (mikelik) [reply src/location.cpp]: merged previous `normalize` and `strip_github_url_to_shorthand` into one function: `strip_github_com`

File: include/antler/project/object.hpp (hunk 1 above)
Comment (dimas1185) [reply include/antler/project/object.hpp]: @mikelik nice investigation, thanks. I'm surprised `map` container internal structures such as `std::pair` are defined in standard. Anyway it doesn't seem to be reliable as we rely on implementation's internal order of initialization. Easier to make it straight than to ponder if it could break or not. This is not performance critical code anyway (although I doubt there is any performance downgrade as object will be moved anyway)

File: include/antler/system/utils.hpp (hunk 4 above)
Comment (dimas1185) [reply include/antler/system/utils.hpp:89]: I doubt that could be a case. if we will have non-critical error output on zero return code that may corrupt parsing of a valid data. Ideally that should be similar to `subprocess.run` in python. i.e. stdout and stderr piping is customizable. but I would leave it to next issue if that will be really needed. If you believe your case is more likely to happen than mine, just leave comment like `pipe error output from console to stdout so caller can parse error string`

Comment (mikelik) [reply include/antler/system/utils.hpp:90]: Sure, done

**Review by dimas1185 — APPROVED** [review]

**Review by jolly-fellow — APPROVED** [review]

**Review by ScottBailey — APPROVED** [review]
